src/layouts/docs/utils/add-full-paths-to-nav-data.ts

```diff
@@ -0,0 +1,48 @@
+import { NavNode } from '@hashicorp/react-docs-sidenav/types'
+import { MenuItem } from 'components/sidebar'
+
+// TODO: export NavBranch and NavLeaf
+// TODO: types from react-docs-sidenav.
+// TODO: isNavBranch & isNavLeaf might
+// TODO: also make sense to include in the
+// TODO: react-docs-sidenav component,
+// TODO: maybe in the types.ts file as well.
+interface NavBranch {
+  title: string
+  routes: NavNode[]
+}
+
+interface NavLeaf {
+  title: string
+  path: string
+}
+
+function isNavBranch(value: NavNode): value is NavBranch {
+  return value.hasOwnProperty('routes')
+}
+
+function isNavLeaf(value: NavNode): value is NavLeaf {
+  return value.hasOwnProperty('path')
+}
+
+function addFullPathsToNavData(
+  nodes: NavNode[],
+  basePaths: string[]
+): MenuItem[] {
+  return nodes.map((n) => addFullPathToNavNode(n, basePaths))
+}
+
+function addFullPathToNavNode(node: NavNode, basePaths: string[]): MenuItem {
+  if (isNavBranch(node)) {
+    // For nodes with routes, add fullPaths to all routes
+    return { ...node, routes: addFullPathsToNavData(node.routes, basePaths) }
+  } else if (isNavLeaf(node)) {
+    // For nodes with paths, add fullPath to the node
+    return { ...node, fullPath: `/${basePaths.join('/')}/${node.path}` }
+  } else {
+    // Otherwise return the node unmodified
+    return node
+  }
+}
```

Comment:
Another topic to add to a follow-up discussion: in the same spirit of these changes, I think we should also generate the menu item `id`s at build time since they're based off of data that's available at that time. Something to be mindful of when we do that: make sure we are only iterating over the menu items once, so we will probably need one function that adds full paths & ids and then some other functions for generating those properties.

Reply:
☝️ yeah this is a great call - I can try to stub something out quickly to add `id`s. 

Is the concern around iterating over menu items only once performance? Or is there something else?

Reply:
Took a shot at this with a single iteration 👍